internal/app/singularity/pull.go

```diff
@@ -0,0 +1,269 @@
+func LibraryPull(name, ref, transport, fullURI, libraryURI, keyServerURL, authToken string, force, unauthenticated bool) error {
+	if !force {
+		if _, err := os.Stat(name); err == nil {
+			return fmt.Errorf("image file already exists: %q - will not overwrite", name)
+		}
+	}
+
+	libraryImage, err := client.GetImage(libraryURI, authToken, fullURI)
+	if err != nil {
+		return fmt.Errorf("while getting image info: %v", err)
+	}
+
+	// required in order to properly allow for library pulls without transport in uri
+	// otherwise uri becomes malformed see https://github.com/sylabs/singularity/pull/2683
+	var imageName string
+	if transport == "" {
+		imageName = uri.GetName("library://" + fullURI)
+	} else {
+		imageName = uri.GetName(fullURI)
+	}
+	imagePath := cache.LibraryImage(libraryImage.Hash, imageName)
+	exists, err := cache.LibraryImageExists(libraryImage.Hash, imageName)
+	if err != nil {
+		return fmt.Errorf("unable to check if %v exists: %v", imagePath, err)
+	}
+	if !exists {
+		sylog.Infof("Downloading library image")
+		if err = client.DownloadImage(imagePath, fullURI, libraryURI, true, authToken); err != nil {
+			return fmt.Errorf("unable to Download Image: %v", err)
+		}
+
+		if cacheFileHash, err := client.ImageHash(imagePath); err != nil {
+			return fmt.Errorf("error getting imagehash: %v", err)
+		} else if cacheFileHash != libraryImage.Hash {
+			return fmt.Errorf("cached file hash(%s) and expected hash(%s) does not match", cacheFileHash, libraryImage.Hash)
+		}
+	}
+
+	// Perms are 777 *prior* to umask
+	dstFile, err := os.OpenFile(name, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0777)
+	if err != nil {
+		return fmt.Errorf("%v", err)
+	}
+	defer dstFile.Close()
+
+	srcFile, err := os.OpenFile(imagePath, os.O_RDONLY, 0444)
+	if err != nil {
+		return fmt.Errorf("%v", err)
+	}
+	defer srcFile.Close()
+
+	// Copy SIF from cache
+	_, err = io.Copy(dstFile, srcFile)
+	if err != nil {
+		return fmt.Errorf("%v", err)
+	}
+
+	var retErr error
+	// check if we pulled from the library, if so; is it signed?
+	if !unauthenticated {
+		imageSigned, err := signing.IsSigned(name, keyServerURL, 0, false, authToken, true)
+		if err != nil {
+			// err will be: "unable to verify container: %v", err
+			sylog.Warningf("%v", err)
+			// if theres a warning, return set error to indicate exit 1
+			retErr = ErrLibraryUnsigned
+		}
+		// if container is not signed, print a warning
+		if !imageSigned {
+			fmt.Fprintf(os.Stderr, "This image is not signed, and thus its contents cannot be verified.\n")
+			resp, err := sypgp.AskQuestion("Do you with to proceed? [N/y] ")
+			if err != nil {
+				return fmt.Errorf("unable to parse input: %v", err)
+			}
+			if resp == "" || resp != "y" && resp != "Y" {
+				fmt.Fprintf(os.Stderr, "Aborting.\n")
+				err := os.Remove(name)
+				if err != nil {
+					return fmt.Errorf("unable to delete the container: %v", err)
+				}
+				// set error to exit with status 10 after replying no
```

Comment:
Would it be possible to improve the comment? I personally do not understand what it really means.